### PR TITLE
feat(ui):feature/lobby/username and button status

### PR DESCRIPTION
**Pull request by enakecanovic**
When the game is opened, the buttons are disabled until the username is typed in. If the username is already defined in a previous game, it will be saved in shared storage and can be used again. If this is the case, the buttons will be automatically enabled.

### CODE REVIEW

**Review by Kirschzelle — COMMENTED** [review]
All in all seems properly implemented changes, everything running etc. you tried a new feature shared preference, so I like where this is going.

Just check comments and try to resolve the conversations.

Would it be possible to limit the letter count of the username, lets say to like 4 or 5 letters for example? Working on the game ui right now this would save me a lot of pain.

Apart from that I would implement avatars (if you want to do them) with a recyclerview, and a shapeable image-view next to your name that shows the chosen avatar. (Sets img source to img source of last recently clicked avatar in recyclerview.) As dominik said Spinner would also work, but I personally would prefer recyclerview from a design perspective. Glad to be told off though if ppl disagree. ^^

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 1]
```diff
@@ -12,10 +16,12 @@ import at.aau.edu.wizards.util.permission.PermissionHandler
 private const val DISCOVER_FRAGMENT_TAG = "DISCOVER_FRAGMENT_TAG"
 private const val LOBBY_FRAGMENT_TAG = "LOBBY_FRAGMENT_TAG"
 private const val GAME_BOARD_FRAGMENT_TAG = "GAME_BOARD_FRAGMENT_TAG"
+private const val SHARED_PREFERENCE_USERNAME_KEY = "USERNAME";
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:19]: unnecessary ;

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 2]
```diff
@@ -12,10 +16,12 @@ import at.aau.edu.wizards.util.permission.PermissionHandler
...
+private const val SHARED_PREFERENCE_NAME = "SHARED_DATA"
 
 class MainActivity : AppCompatActivity() {
 
-    private lateinit var binding: ActivityMainBinding
+    private lateinit var binding: ActivityMainBinding;
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:24]: again unnecessary change

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 3]
```diff
@@ -43,15 +49,50 @@ class MainActivity : AppCompatActivity() {
     }
 
     private fun setupUi() {
+        val sharedPrefs: SharedPreferences = getSharedPreferences(getString(R.string.shared_storage), Context.MODE_PRIVATE)
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:52]: do we need SharedPreferences? I'm new to this aswell, but as far as i have red this uses a lot of resources and is used to communicate between activities. We only use one activity though, so is there a reason you decided to implement this instead of just saving it locally inside the activity?

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 4]
```diff
@@ -43,15 +49,50 @@ class MainActivity : AppCompatActivity() {
...
+                }
+            }
+
+            override fun afterTextChanged(editable: Editable) {}
+        })
         binding.clientButton.setOnClickListener {
-            showDiscoverFragment()
+            if (binding.clientButton.isEnabled) {
+                showDiscoverFragment()
+            }
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:77]: here i would also add an else that for example shows a toast "please enter a username" when the user clicks on the greyed out button.

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 5]
```diff
@@ -43,15 +49,50 @@ class MainActivity : AppCompatActivity() {
...
         }
 
         binding.serverButton.setOnClickListener {
-            showLobby()
+            if (binding.serverButton.isEnabled) {
+                showLobby();
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:82]: .;.;.

File: app/src/main/res/layout/activity_main.xml [hunk 6]
```diff
@@ -27,6 +27,16 @@
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintTop_toBottomOf="@id/server_button" />
 
+    <EditText
```
Comment (Kirschzelle) [inline app/src/main/res/layout/activity_main.xml:30]: Missing `autofillHints` attribute - just add it:

https://developer.android.com/reference/androidx/autofill/HintConstants#AUTOFILL_HINT_USERNAME()

File: app/src/main/res/layout/activity_main.xml [hunk 7]
```diff
@@ -27,6 +27,16 @@
...
+        android:id="@+id/input_username"
+        android:layout_height="wrap_content"
+        android:layout_width="match_parent"
+        android:inputType="text"
+        android:hint="Enter username"
```
Comment (Kirschzelle) [inline app/src/main/res/layout/activity_main.xml:35]: this is fine for testing and getting things running (so its basically fine for now) but if you put this constant string in strings.xml we get rid of a smell

dominik might not see this as important though since he did that a lot ^^

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 8]
```diff
@@ -43,15 +49,50 @@ class MainActivity : AppCompatActivity() {
...
+        if (usernameAlreadyExists(sharedPrefs)) {
+            binding.inputUsername.setText(sharedPrefs.getString(SHARED_PREFERENCE_USERNAME_KEY, ""))
+            setButtonsEnableStatus(true)
+        }
+        else {
+            setButtonsEnableStatus(false)
+        }
+
+        binding.inputUsername.addTextChangedListener(object : TextWatcher {
+            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:62]: empty function declaration, add a nested comment explaining why it is empty. (this is a smell, i understand that you just want this to do nothing and need to implement it because class is abstract)

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 9]
```diff
@@ -43,15 +49,50 @@ class MainActivity : AppCompatActivity() {
...
+            override fun onTextChanged(typedText: CharSequence, i: Int, i1: Int, i2: Int) {
+                if (typedText.isNotEmpty()) {
+                    setButtonsEnableStatus(true)
+                }
+                else {
+                    setButtonsEnableStatus(false)
+                }
+            }
+
+            override fun afterTextChanged(editable: Editable) {}
```
Comment (Kirschzelle) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:72]: same, add nested comment

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt (hunk 4 above)
Comment (lechld) [reply app/src/main/java/at/aau/edu/wizards/MainActivity.kt:77]: Code not 100% accurate, but you can also show Snackbar instead. Error case handling is important, always try to keep in mind that users can be "really stupid"

`Snackbar.makeText(context, R.string.empty_username_warning, Snackbar.LengthLong).show()`

File: app/src/main/res/layout/activity_main.xml (hunk 7 above)
Comment (lechld) [reply app/src/main/res/layout/activity_main.xml:35]: Yeah u got me. It's bad practice to use hardcoded strings (ide warns for a reason). I thought as long as I don't know what text we should really display....
Sonar gives us the tool that we can replace all of them in one step. With all languages we know.
@enakecanovic I think you are capable of an extra language, maybe we should think about adding that translation as well?

**Review by lechld — APPROVED** [review]

File: app/src/main/res/layout/activity_main.xml (hunk 7 above)
Comment (Kirschzelle) [reply app/src/main/res/layout/activity_main.xml:35]: I really like the idea of putting an extra language there, free code lines aswell btw.

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt (hunk 3 above)
Comment (Kirschzelle) [reply app/src/main/java/at/aau/edu/wizards/MainActivity.kt:52]: Yes, but as far as I can see if the application as a whole gets killed we still loose the data and since there is only one activity in our application there won't be a case where we it will be killed. But I guess we just future proof with it?

Comment (lechld) [reply app/src/main/java/at/aau/edu/wizards/MainActivity.kt:52]: Activity can be killed any time, you can get a phone call and the systems kills our Activity because low memory. But we handle such restoration cases not properly anyhow. Here it's more about avoiding entering name when opening the app second (third,...) time

**Comment by Kirschzelle** [comment]
Please solve merge conflicts, additionally we should pass val of names into the player models in the game (so into the start of the game) - we can set this as goal for next week though, but merge conflicts need to be resolved (merge current master into your branch)

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt (hunk 3 above)
Comment (enakecanovic) [reply app/src/main/java/at/aau/edu/wizards/MainActivity.kt:52]: as discussed on DD,  we stick to using SharedPrefs until we find a better solution

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt (hunk 4 above)
Comment (enakecanovic) [reply app/src/main/java/at/aau/edu/wizards/MainActivity.kt:77]: As the button is disabled until the username is given, the onClick event does not fire, therefore, we cannot detect a click on the disabled button. We can do this by displaying a label to inform the user that they first need to enter username to proceed.

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1,7 +1,10 @@
 <resources>
     <string name="app_name">Wizards</string>
     <string name="start_game">Start game</string>
-    <string name="behave_as_server">Create lobbby</string>
+    <string name="enter_username">Enter username</string>
+    <string name="autofill_hint_username">username</string>
+    <string name="shared_storage">Shared storage</string>
```
Comment (lechld) [inline app/src/main/res/values/strings.xml:6]: the key of the storage should not be localized!

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 11]
```diff
@@ -14,16 +18,22 @@ import at.aau.edu.wizards.util.permission.PermissionHandler
 private const val DISCOVER_FRAGMENT_TAG = "DISCOVER_FRAGMENT_TAG"
 private const val LOBBY_FRAGMENT_TAG = "LOBBY_FRAGMENT_TAG"
 private const val GAME_BOARD_FRAGMENT_TAG = "GAME_BOARD_FRAGMENT_TAG"
+const val SHARED_PREFERENCE_USERNAME_KEY = "USERNAME"
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:21]: that constant is only used in `MainViewModel`. Please move it there and change visibility to private.

File: app/src/main/java/at/aau/edu/wizards/MainActivity.kt [hunk 12]
```diff
@@ -60,15 +70,54 @@ class MainActivity : AppCompatActivity() {
...
+
+            override fun afterTextChanged(editable: Editable) {
+                // Not required
+            }
+        })
+
         binding.clientButton.setOnClickListener {
-            showDiscoverFragment()
+            if (binding.clientButton.isEnabled) {
+                val username = binding.inputUsername.text.toString()
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/MainActivity.kt:101]: here and below we have a warning, usually that can be fixed giving a different name. Eg `newUserName`.

File: app/src/main/java/at/aau/edu/wizards/MainViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package at.aau.edu.wizards
+
+import android.content.SharedPreferences
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+
+class MainViewModel constructor(private val sharedPreferences: SharedPreferences) : ViewModel() {
```
Comment (lechld) [inline app/src/main/java/at/aau/edu/wizards/MainViewModel.kt:7]: writing that `constructor` is unnecessary :)

File: app/src/main/res/layout/activity_main.xml [hunk 14]
```diff
@@ -35,33 +37,62 @@
...
 
     <Button
         android:id="@+id/client_button"
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
-        android:layout_margin="20dp"
+        android:layout_margin="10dp"
+        android:layout_marginBottom="0dp"
         android:fontFamily="@font/bitter"
+        android:rotationY="0"
```
Comment (lechld) [inline app/src/main/res/layout/activity_main.xml:56]: why do we add `rotationY="0"`

File: app/src/main/res/layout/activity_main.xml [hunk 15]
```diff
@@ -35,33 +37,62 @@
...
+        android:orientation="horizontal"/>
+
+    <EditText
+        android:id="@+id/input_username"
+        android:layout_width="0dp"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="-20dp"
+        android:hint="@string/enter_username"
+        android:inputType="text"
+        android:maxLength="5"
```
Comment (lechld) [inline app/src/main/res/layout/activity_main.xml:77]: Does `5` really make sense? Seems a bit short to me

File: app/src/main/res/layout/activity_main.xml [hunk 16]
```diff
@@ -35,33 +37,62 @@
...
+        android:name="at.aau.edu.wizards.ui.lobby.LobbyFragment"
         android:layout_width="0dp"
         android:layout_height="0dp"
         app:layout_constraintBottom_toBottomOf="parent"
         app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintHorizontal_bias="0.0"
         app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_bias="0.0" />
```
Comment (lechld) [inline app/src/main/res/layout/activity_main.xml:96]: we talked about that, please remove and check if the same is applied somewhere else in that file.

**Review by lechld — COMMENTED** [review]
Open topics:

* Unit test for MainViewModel
* Code isn't properly formatted again. https://discord.com/channels/1090224949937057792/1090227413935468674/1106289576240218244

**Comment by lechld** [comment]
Good work to introduce ViewModel. That's exactly how I wanted it to be.

UI has some issues! We have full width `TextView` where we can enter only five characters. Also the warning we agreed on is missing

File: app/src/main/res/layout/activity_main.xml (hunk 15 above)
Comment (Kirschzelle) [reply app/src/main/res/layout/activity_main.xml:77]: Was an example length provided by me - we want to avoid new lines breaking up the name,
Ex.:
Lech
  ld
Instead of:
Lechld
and there is not that much space to write the name without a newline in the current gameboard. Exact number of characters would have to be tested - but 4-5 should work.

File: app/src/test/java/at/aau/edu/wizards/MainViewModelTests.kt [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+    @Test
+    fun getUsernameTests_testExactString_returnsExactString() {
+        val expectedString = "some random string"
+        whenever(sharedPrefs.getString(SHARED_PREFERENCE_USERNAME_KEY, null)).doReturn(expectedString)
+        assertEquals(expectedString, mainViewModel.getUsername())
+    }
+
+    @Test
+    fun saveUsernameTests_setUsername_accepted() {
+        val editor = createEditor()
```
Comment (lechld) [inline app/src/test/java/at/aau/edu/wizards/MainViewModelTests.kt:35]: you can use a mock for the editor too, way simpler to write

```
@Test
fun saveUsernameTests_setUsername_accepted() {
        val editor = mock<SharedPreferences.Editor>()
        val username = "some username here"

        whenever(sharedPrefs.edit()).doReturn(editor)
        whenever(editor.putString(SHARED_PREFERENCE_USERNAME_KEY, username)).doReturn(editor)

        mainViewModel.saveUsername(username)

        then(sharedPrefs).should().edit()
        then(editor).should().putString(SHARED_PREFERENCE_USERNAME_KEY, username)
    }
```

File: app/src/test/java/at/aau/edu/wizards/MainViewModelTests.kt [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+import android.content.SharedPreferences
+import org.junit.Assert.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.Test
+import org.mockito.Mockito
+import org.mockito.kotlin.*
+
+class MainViewModelTests {
+    private val sharedPrefs = Mockito.mock(SharedPreferences::class.java, Mockito.RETURNS_DEEP_STUBS)
```
Comment (lechld) [inline app/src/test/java/at/aau/edu/wizards/MainViewModelTests.kt:11]: simpler to write, same effect:
`private val sharedPrefs = mock<SharedPreferences>()`

**Review by lechld — COMMENTED** [review]
Just left some feedback which can help with other tests in future, you can adjust to that or ignore it, nevermind

**Review by lechld — COMMENTED** [review]
Just left some feedback which can help with other tests in future, you can adjust to that or ignore it, nevermind
